Improve scheduler performance by reducing lock contention

psalz opened the pull request:
This PR is based on #109 so that one should be merged first.

This improves the performance of the scheduler in two ways:
  - Address a long standing issue (we've had a TODO note about this for years...) where the scheduler thread would hold the lock on the `events` queue until a task was fully built and flushed, effectively causing it to operate in lockstep with the main thread. This has been fixed by double-buffering the event queue, with the scheduler thread only locking the mutex to swap the two.
  - Reduce contention around the `task_manager`'s mutex by passing a task pointer to the scheduler directly, instead of it repeatedly calling `task_manager::get_task` (while the main thread is trying to acquire the lock for producing new tasks).

Additionally I've reduced the number of tasks generated in the "soup" topology benchmark so that the timings are in about the same order of magnitude as for the other topologies (= fits better onto a chart).

Here is a comparison between the original "lockstep" scheduler and the double-buffered "decoupled" one, but still with frequent calls to `get_task`. There are two bars for each benchmark in each group (same colors), with the right bar representing the decoupled scheduler. The two "ST *" groups refer to the single-threaded baseline implementation that generates commands in the main thread, circumventing the scheduler altogether (numbers in these groups should not be affected by the changes).

![image](https://user-images.githubusercontent.com/791348/160912323-a8e370e3-acad-40a6-8332-e3498ef60a35.png)

We can see that this is an easy win overall, with numbers improving across the board, in particular for the "soup" and "wave_sim" topologies. Additionally jitter is greatly reduced.

Unfortunately numbers are not *quite* as clear for the second change. Here we now compare the decoupled scheduler against the task pointer being passed around (the left bar here is the right bar above):

![image](https://user-images.githubusercontent.com/791348/160912776-4e537762-e51f-4692-aa26-6b8214a28407.png)

There are again considerable improvements for "soup" and "wave_sim", which now also outperform the single-threaded variant in most cases. Jitter is again reduced across the board. However with 4 nodes, there are some numbers that are worse, in particular for the "Immediate" group. While I was able to replicate this several times, I've also encountered an instance where performance was better everywhere, with the "Immediate" group outperforming "ST Immediate" for most topologies. I therefore attribute this "worsening" to overall jitter in our benchmarking setup, and given the significance of the "soup" and "wave_sim" results, I would still consider this an improvement overall.

fknorr commented on include/scheduler.h:
```diff
@@ -51,13 +51,16 @@ namespace detail {
 
-		std::queue<scheduler_event> events;
+		std::queue<scheduler_event> available_events;
+		std::queue<scheduler_event> in_flight_events;
```
`in_flight_events` is only used inside `schedule()`, so it could be a local variable. This would make it more obvious that it is not synchronized by `events_mutex`.

psalz replied:
Good idea, done!

fknorr commented on src/scheduler.cc:
```diff
@@ -55,6 +65,7 @@ namespace detail {
 	void scheduler::shutdown() {
 		abstract_scheduler::shutdown();
 		if(worker_thread.joinable()) { worker_thread.join(); }
+		assert(available_events.empty() && in_flight_count == 0);
```
This is missing a lock on `events_mutex` (can potentially race with `notify()`). It should be enough to take it in a `#ifndef NDEBUG' though.

psalz replied:
Right, although this is moot now, since the refactoring into `abstract_scheduler` doesn't allow me to access these members anyway. I've removed the assertion.

fknorr commented on include/graph_generator.h:
```diff
@@ -77,18 +77,16 @@ namespace detail {
 
 		// Build the commands for a single task
-		void build_task(task_id tid, const std::vector<graph_transformer*>& transformers);
+		void build_task(const task& tsk, const std::vector<graph_transformer*>& transformers);
```
I like the use of `task&` instead of `task*` for non-null parameters here. IMO this should be policy for all non-null `task` and `abstract_command` parameters, which are currently pointers everywhere else for some reason.

psalz replied:
Agreed!

fknorr commented on include/scheduler.h:
```diff
@@ -51,13 +51,16 @@ namespace detail {
-		std::queue<scheduler_event> events;
+		std::queue<scheduler_event> available_events;
+		std::queue<scheduler_event> in_flight_events;
+		std::atomic_size_t in_flight_count = 0;
```
This is only really needed by `scheduler::is_idle()`, which will be removed by #86 (only use case was the ad-hoc `slow_full_sync`). Depending on merge order, we can do without this variable entirely.

fknorr commented on src/scheduler.cc:
```diff
@@ -14,35 +14,45 @@ namespace detail {
+			}
+
+			while(!in_flight_events.empty()) {
+				const auto event = std::move(in_flight_events.front());
```
Clang-Tidy: `std::move` of the expression of the trivially-copyable type `__gnu_cxx::__alloc_traits<std::allocator<celerity::detail::scheduler_event>, celerity::detail::scheduler_event>::value_type` (aka `celerity::detail::scheduler_event`) has no effect; remove `std::move()`

psalz replied:
I got the same diagnostic and deliberately kept this so that if `scheduler_event` would for some reason no longer be trivially-copyable in the future, we're already set. As it says, it's a no-op right now so it doesn't really do any harm either, IMO.

fknorr replied:
Fine with me, but we could still document the intent with a `// NOLINT(performance-move-const-arg)`.

fknorr commented on src/scheduler.cc:
```diff
@@ -14,35 +14,45 @@ namespace detail {
+					return;
+				}
+
+				if(event.type == scheduler_event_type::TASK_AVAILABLE) {
```
This could be an assertion instead of an if, which would make it more obvious that no event is skipped – and by extension `in_flight_count` never misses an update.

psalz replied:
Right, changed.

fknorr reviewed (commented):
Great improvement, thanks! Even if there is not always a benefit to the single-threaded setup, this is an important requirement for any parallel CDAG generation we might want to do in the future.

Could you add another bar chart for the TDAG times as a baseline to see what the best-case improvement of the separate scheduler thread would be?

psalz commented:
> Could you add another bar chart for the TDAG times as a baseline to see what the best-case improvement of the separate scheduler thread would be?

Yes! I've plotted the single-threaded TDAG+CDAG times against the dedicated scheduler thread. If we had perfect overlapping, the scheduler times would be equal to the single-threaded CDAG times. For 4 nodes we are not that far off!

![image](https://user-images.githubusercontent.com/791348/161047484-b188aae7-3780-4be1-8e09-a12e68fa5542.png)

fknorr commented:
Thanks! LGTM up to the `in_flight_count` remark. Some notes for future investigation into scheduler-thread performance:

It's remarkable how far results differ between topologies. It seems intuitive that the benefit should be maximal when TDAG and CDAG generation are most similar in execution time (`soup` in both graphs, `wave_sim` in the single-node case). I assume that the positive result for `wave_sim` in the 4-node example is simply due to jitter, because in no other scenario can the TDAG generation latency be hidden completely.

To me it seems like the 1-node soup should be the perfect condition for parallelization since TDAG and CDAG can be almost fully overlapped and the total runtime is long enough to avoid any constant overhead that might be visible in faster runs (e.g. `chain`). Yet we only save about 23% instead of the theoretical 45%. My guess is that since each individual task is very quick to compute, the scheduler has to block on its queue after every task, so we add an OS scheduler round-trip to each iteration.

Also it is notable that for 1-node soup, TDAG generation is more expensive than CDAG. Both generate mostly identical graphs, but the CDAG is also serialized (and contains the infamous `debug_label`), so I would expect it to be the other way round. Possible causes that come to mind:
- Allocations for command group type erasure and range mappers
- `task_mutex`, locked and unlocked for every task – but it is not contended here, so locking should never need a syscall

psalz commented:
I've rebased onto master, which now includes the task ring buffer. Due to this the second optimization (passing around a task-pointer) is moot, as `task_manager::get_task` is now lock-free. Nevertheless, I think it's still a good thing to do.

With the ring buffer times are lower in general and the results after decoupling the scheduler are a lot less dramatic. Improvements are most noticeable for tasks where computing the command graph isn't cheap, e.g. wave_sim.

![image](https://user-images.githubusercontent.com/791348/179746849-f0fea509-2b5c-4fda-b512-7d5a46ec5d87.png)

![image](https://user-images.githubusercontent.com/791348/179747014-9f5f9d25-892c-4617-8d76-defe75f5000d.png)

Note besides decreasing the number of tasks for the "soup" benchmark (as described above), I've now also changed the benchmarks to use the task-manager callback to notify the scheduler of newly created tasks. This has the effect that now horizons will also be built by the benchmark (previously they were silently ignored).

fknorr reviewed (approved):
LGTM! Some day we should do a pass over both task and command graph code replacing DAG node pointers with references where they are guaranteed to be non-null.

PeterTh commented on include/scheduler.h:
```diff
@@ -1,23 +1,23 @@
 #pragma once
 
+#include <atomic>
```
I assume this is needed now because it was transitive over "types.h" before?

psalz replied:
Actually this was only needed before the rebase onto epochs - removed.

PeterTh commented on include/task_manager.h:
```diff
@@ -74,6 +75,7 @@ namespace detail {
 				task& task_ref = register_task_internal(std::move(reservation), std::move(cgh).into_task());
+				tsk_ptr = &task_ref;
 
 				compute_dependencies(tid);
```
Maybe this should also directly take tasks? The other call-site could do this as well (if `reduce_execution_front` return a task rather than an id, which it probably should).

psalz replied:
Right, I've changed the task manager to also pass around references instead of ids in most places.

PeterTh reviewed (approved):
LGTM, other than the comment on going a bit farther with the task_id -> task refactoring (could also be a separate PR).

PeterTh commented on src/task_manager.cc:
```diff
@@ -237,12 +236,12 @@ namespace detail {
 
 		// it's important that we don't hold the lock while doing this
-		invoke_callbacks(tid);
+		invoke_callbacks(get_task(tid));
```
Could get rid of this `get_task` by storing the task returned by `reduce_execution_front`.

psalz replied:
Done!

PeterTh commented on src/task_manager.cc:
```diff
@@ -253,15 +252,15 @@ namespace detail {
 
 		// it's important that we don't hold the lock while doing this
-		invoke_callbacks(tid);
+		invoke_callbacks(get_task(tid));
```
Same here.

psalz replied:
Done!
